backend/packages/Upgrade/src/api/services/ExperimentAssignmentService.ts

```diff
@@ -248,25 +210,38 @@ export class ExperimentAssignmentService {
     }
 
     // experiment level inclusion and exclusion
-    const [filteredExperiments, exclusionReason] = await this.experimentLevelExclusionInclusion(
-      globalFilteredExperiments,
-      experimentUser
-    );
-
-    if (filteredExperiments.length) {
-      let monitoredDocument: MonitoredDecisionPoint = await this.monitoredDecisionPointRepository.findOne({
+    const [, exclusionReason] = await this.experimentLevelExclusionInclusion(globalFilteredExperiments, userDoc);
+    let monitoredDocument: MonitoredDecisionPoint;
+    if (experimentId && experiments.length) {
+      monitoredDocument = await this.monitoredDecisionPointRepository.findOne({
         where: {
           site: site,
           target: target,
           user: userId,
         },
         relations: ['user'],
       });
+      const selectedExperimentDP = dpExperiments.find((dp) => dp.experiment.id === experimentId);
+      const experiment = experiments[0];
+      const { conditions } = experiment;
+      const payloadCondition = conditions.flatMap((condition) => condition.conditionPayloads);
+
+      const matchedCondition = conditions.filter((dbCondition) => dbCondition.conditionCode === condition);
+      const matchedPayloadCondition = payloadCondition.filter((con) => con.payloadValue === condition);
+      if (matchedCondition.length === 0 && matchedPayloadCondition.length === 0 && condition !== null) {
+        const error = new Error(`Condition not found: ${condition}`);
+        (error as any).type = SERVER_ERROR.CONDITION_NOT_FOUND;
+        logger.error(error);
+        throw error;
+      }
 
-      if (experimentId && experiments.length > 0) {
-        const selectedExperimentDP = dpExperiments.find((dp) => dp.experiment.id === experimentId);
+      if (
+        (experiment.state === EXPERIMENT_STATE.ENROLLING ||
+          experiment.state === EXPERIMENT_STATE.ENROLLMENT_COMPLETE) &&
+        !previewUser
+      ) {
+        const experiment = await this.experimentService.findOne(experimentId);
```

Comment:
Why are you getting the experiment document from DB? Don't you already have that information?

Reply:
Yeah, this code was already existing, and was fetching the complete experiment document, as the earlier experiment object didnt had partitions and conditionPayloads keys. Partitions key was needed for checking enrollmentComplete condition, so added that in the relations for the earlier available experiment document. So, now I have removed this excess fetching of redundant data.